gemd/util/impl.py

```diff
@@ -98,34 +109,118 @@ def _substitute(thing: Any,
+                     for k, v in replacement.as_dict().items()}
+        new = replacement.build(new_attrs)
     else:
-        new = thing
+        new = replacement
 
     if thing.__hash__ is not None:
         visited[thing] = new
-    if new.__hash__ is not None:
-        visited[new] = new
 
-    # assert type(thing) == type(new), "{} is not {}".format(type(thing), type(new))
     return new
 
 
+def _substitute_inplace(thing: Any,
+                        sub: Callable[[object], object],
+                        applies: Callable[[object], bool],
+                        visited: Dict[object, object] = None) -> object:
+    """
+    Generic recursive in-place substitute function.
+
+    Iteratively crawls the passed structure, substituting elements with sub(element) when
+    applies(element) is true and the element is mutable.
+
+    Parameters
+    ----------
+    thing: Any
+        The object to traverse with substitution.
+    sub: Callable[[object], object]
+        Function which provides substitute for value; should not have side-effects.
+    applies: Callable[[object], bool]
+        Function which defines the domain for the sub function to be invoked.
+
+    """
+    def _key(obj):
+        if obj.__hash__ is not None:
+            return obj
+        else:
+            return id(obj)
+
+    orig_key = _key(thing)
+    if visited is None:
+        visited = {}
+    if orig_key in visited:
+        return visited[orig_key]
+
+    if applies(thing):
+        thing = sub(thing)
+    visited[orig_key] = thing  # Store before we start recursing
+
+    if _cached_isinstance(thing, list):  # Change list in place
+        for i, x in enumerate(thing):
+            thing[i] = _substitute_inplace(x, sub, applies, visited)
+    elif _cached_isinstance(thing, tuple):  # Tuples are immutable; regenerate
+        thing = tuple(_substitute_inplace(x, sub, applies, visited) for x in thing)
+        visited[orig_key] = thing  # We mutated it
+    elif _cached_isinstance(thing, dict):  # Change dict in place, both keys & values
+        remove = set()  # Store todos because can't mutate a dict in a loop
+        update = dict()
+        for k, v in thing.items():
+            new_k = _substitute_inplace(k, sub, applies, visited)
+            new_v = _substitute_inplace(v, sub, applies, visited)
+            if id(k) != id(new_k):
+                remove.add(k)
+                update[new_k] = v
+            if id(v) != id(new_v):
+                update[new_k] = new_v
+        for k in remove:
+            thing.pop(k, None)
+        thing.update(update)
+    elif _cached_isinstance(thing, DictSerializable):
+        for k, v in thing.as_dict().items():  # Assume key can't change b/c it's an attribute
+            new_v = _substitute_inplace(v, sub, applies, visited)
+            if id(v) != id(new_v):
+                _setter_by_name(type(thing), k)(thing, new_v)
+
+    return thing
+
+
+@functools.lru_cache(maxsize=None)
```

Comment:
This applies here and elsewhere in the PR.

I think that `@functools.lru_cache(maxsize=None)` is equivalent to just `@functools.cache`. Should we just switch to that in all places?

Reply:
That was introduced in 3.9.
https://docs.python.org/3.10/library/functools.html#functools.cache